data/shared/citizen/scripting/lua/scheduler.lua

```diff
@@ -280,23 +275,27 @@ Citizen.SetEventRoutine(function(eventName, eventPayload, eventSource)
 
 	-- try finding an event handler for the event
 	local eventHandlerEntry = eventHandlers[eventName]
-	
-	-- deserialize the event structure (so that we end up adding references to delete later on)
-	local data = msgpack.unpack(eventPayload)
 
 	if eventHandlerEntry and eventHandlerEntry.handlers then
 		-- if this is a net event and we don't allow this event to be triggered from the network, return
 		if eventSource:sub(1, 3) == 'net' then
-			if not eventHandlerEntry.safeForNet and not alwaysSafeEvents[eventName] then
+			if not eventHandlerEntry.safeForNet then
 				Citizen.Trace('event ' .. eventName .. " was not safe for net\n")
 
+				_G.source = lastSource
 				return
 			end
 
 			deserializingNetEvent = { source = eventSource }
 			_G.source = tonumber(eventSource:sub(5))
+		elseif IsDuplicityVersion() and eventSource:sub(1, 12) == 'internal-net' then
+			deserializingNetEvent = { source = eventSource:sub(10) }
+			_G.source = tonumber(eventSource:sub(14))
 		end
 
+		-- deserialize the event structure (so that we end up adding references to delete later on)
```

Comment:
Moving this seems undesirable with regards to the precarious case of broadcast funcrefs: every reference _submitted_ must be _deserialized_ in order to be GC'd at _some_ point. A more viable approach would be to annotate the ScRT event call with a set of funcref IDs to immediately discard upon ignoring an event, and _then_ one can move serialization to be only on-demand.

This was moved there in [b250df1cf1859e21d8a2078b23ade14e098a6ebd](https://github.com/citizenfx/fivem/commit/b250df1cf1859e21d8a2078b23ade14e098a6ebd#diff-27f3d1b4362eb7a4837bdb80d6a1a71c08e408e134ff43adae7116b4632c6466R146), and this comment was amended to indicate that it was up there for a reason.

Current broadcast funcrefs might be extremely harmful, for they will lead to extremely confusing GC scenarios if even _one_ ScRT hangs on to them for too long, but this change will lead to guaranteed memory leaks... I _think_. **Not sure in fact, would need explicit testing and hard to submarine in here.**

I do think the comment could be clearer in this regard!

Reply:
I think the best thing for *this* PR is just to revert the change and create a seperate PR for this specifically. In hindsight, I was overstepping the scope of this PR with this change and should have just changed what needed to be changed to get the feature to work correctly.

Happy to move forward with that or should I include such changes here too?

Reply:
Would indeed be suggested to put any such changes elsewhere, yes.